Src/serial_interface.c

```diff
@@ -1,456 +1,460 @@
-				// next 4 bytes are transition frame number (uint32_t)
-				instance.controller_data_bytes_read = 0;
-				instance.state = SERIAL_TRANSITION_3;  // intentional fall through
-			case SERIAL_TRANSITION_3:
-				instance.controller_data_buffer[instance.controller_data_bytes_read++] = input;
-
-				if(instance.controller_data_bytes_read < sizeof(uint32_t)) // only save 4 bytes
-				{
-					break;
-				}
-
-				uint32_t tempVal;
-				memcpy(&tempVal, instance.controller_data_buffer, sizeof(tempVal));
-
-				// now make a decision based off of the 3rd byte noted earlier
-				if(instance.transition_type == 'A') // transition to ACE
-				{
-					if(!AddTransition(TRANSITION_ACE, tempVal)) // try adding transition
-					{
-						// adding transition failed
-						instance.state = SERIAL_COMPLETE;
-						serial_interface_output((uint8_t*)"\xFB", 1);
-						break;
-					}
-				}
-				else if(instance.transition_type == 'N')
-				{
-					if(!AddTransition(TRANSITION_NORMAL, tempVal)) // try adding transition
-					{
-						// adding transition failed
-						instance.state = SERIAL_COMPLETE;
-						serial_interface_output((uint8_t*)"\xFB", 1);
-						break;
-					}
-				}
-				else if(instance.transition_type == 'S')
-				{
-					if(!AddTransition(TRANSITION_RESET_SOFT, tempVal)) // try adding transition
-					{
-						// adding transition failed
-						instance.state = SERIAL_COMPLETE;
-						serial_interface_output((uint8_t*)"\xFB", 1);
-						break;
-					}
-				}
-				else if(instance.transition_type == 'H')
-				{
-					if(!AddTransition(TRANSITION_RESET_HARD, tempVal)) // try adding transition
-					{
-						// adding transition failed
-						instance.state = SERIAL_COMPLETE;
-						serial_interface_output((uint8_t*)"\xFB", 1);
-						break;
-					}
-				}
-
-				instance.state = SERIAL_COMPLETE;
-				break;
-			default:
-				break;
-		}
-	}
-
-}
+#include "serial_interface.h"
+
+#include "stm32f4xx_it.h"
+#include "main.h"
+#include "TASRun.h"
+#include "info.h"
+
+#include <string.h>
+#include <stdlib.h>
+
+// only instance of this, but make callers use access functions
+static serial_interface_state_t instance;
+extern RunDataArray *dataptr;
+static uint8_t NullOutputFunction(uint8_t *buffer, uint16_t n)
+{
+	return 0;
+}
+
+void serial_interface_reset() {
+	instance.state = SERIAL_COMPLETE;
+	serial_interface_set_output_function(NullOutputFunction);
+}
+
+void serial_interface_set_output_function(OutputFunction func)
+{
+	instance.output_func = func;
+}
+
+inline uint8_t serial_interface_output(uint8_t *buffer, uint16_t n)
+{
+	return instance.output_func(buffer, n);
+}
+
+void serial_interface_consume(uint8_t *buffer, uint32_t n)
+{
+	for (uint32_t i = 0; i != n; ++i)
+	{
+		uint8_t input = buffer[i];
+		switch(instance.state)
+		{
+			case SERIAL_COMPLETE: // in case more than 1 command is sent at a time
+			case SERIAL_PREFIX:
+				switch(input)
+				{
+					case 'I': // return device info
+						serial_write_InfoBlob();
+            break;
```

Comment:
Unsure why this "break" does not appear to be aligned with any other lines. Tabs vs. spaces? Also, why is the whole file a diff? :(

Reply:
I'd throw money on line-ending conversion

Reply:
It looks like commit 59d5682 is the one that changed the entire file. Would you be willing to revert and then try that merge again? Or merge that manually with a commit that doesn't redo the whole file? this way the git diff remains clean and readable. Thanks!

Reply:
Apparently I had write access to your branch. So I fixed it myself and pushed it. Indeed, @kevans91 was correct. That one file had its windows-style line-endings converted to unix. Might want to remove all "\r" at some point... but in a separate PR.